app/init/resources.php

```diff
@@ -918,38 +918,52 @@ function getDevice(string $root, string $connection = ''): Device
-        if (empty($tokenId) || empty($secret)) {
+        if (empty($tokenId)) {
             return new Document([]);
         }
 
         $token = Authorization::skip(fn () => $dbForProject->getDocument('resourceTokens', $tokenId));
 
-        if ($token->isEmpty() || $token->getAttribute('secret') !== $secret) {
+        if ($token->isEmpty()) {
             return new Document([]);
         }
 
-        if ($token->getAttribute('resourceType') === TOKENS_RESOURCE_TYPE_FILES) {
-            $internalIds = explode(':', $token->getAttribute('resourceInternalId'));
-            $ids = explode(':', $token->getAttribute('resourceId'));
+        $expiry = $token->getAttribute('expire');
 
-            if (count($internalIds) !== 2 || count($ids) !== 2) {
+        if ($expiry !== null) {
+            $now = new \DateTime();
+            $expiryDate = new \DateTime($expiry);
+
+            if ($expiryDate < $now) {
                 return new Document([]);
             }
+        }
 
-            $accessedAt = $token->getAttribute('accessedAt', 0);
-            if (empty($accessedAt) || DatabaseDateTime::formatTz(DatabaseDateTime::addSeconds(new \DateTime(), - APP_RESOURCE_TOKEN_ACCESS)) > $accessedAt) {
-                $token->setAttribute('accessedAt', DatabaseDateTime::now());
-                Authorization::skip(fn () => $dbForProject->updateDocument('resourceTokens', $token->getId(), $token));
-            }
+        return match ($token->getAttribute('resourceType')) {
+            TOKENS_RESOURCE_TYPE_FILES => (function () use ($token, $dbForProject) {
+                $internalIds = explode(':', $token->getAttribute('resourceInternalId'));
+                $ids = explode(':', $token->getAttribute('resourceId'));
 
-            return new Document([
-                'bucketId' => $ids[0],
-                'fileId' => $ids[1],
-                'bucketInternalId' => $internalIds[0],
-                'fileInternalId' => $internalIds[1],
-            ]);
-        }
+                if (count($internalIds) !== 2 || count($ids) !== 2) {
+                    return new Document([]);
+                }
+
+                $accessedAt = $token->getAttribute('accessedAt', 0);
+                if (empty($accessedAt) || DatabaseDateTime::formatTz(DatabaseDateTime::addSeconds(new \DateTime(), - APP_RESOURCE_TOKEN_ACCESS)) > $accessedAt) {
+                    $token->setAttribute('accessedAt', DatabaseDateTime::now());
+                    Authorization::skip(fn () => $dbForProject->updateDocument('resourceTokens', $token->getId(), $token));
+                }
+
+                return new Document([
+                    'bucketId' => $ids[0],
+                    'fileId' => $ids[1],
+                    'bucketInternalId' => $internalIds[0],
+                    'fileInternalId' => $internalIds[1],
+                ]);
+            })(),
+
+            default => throw new Exception(Exception::TOKEN_RESOURCE_TYPE_INVALID),
```

Comment:
should we throw error here, or return empty document like in other cases?

Reply:
I think it makes sense to keep it a bit stricter. add case by case and throw an error.